System Check Script

yashika-batra opened the pull request:
- deleted system check script folder from unit tests
- added system-check to application
- removed lib folder (will use the lib created from drivers instead)
- removed tests/groundstation folder (this is in the feather groundstation repository now)

jbwillis commented on state_machine/applications/system-check/main.py:
```diff
@@ -0,0 +1,115 @@
+run_sun_sensor = False
+run_coil_driver = False
+run_burnwire = False
+solar_boards = input("Are the solar boards attached to the mainboard? (Y/N): ")
```
I don't think we actually need to ask this question. I'd rather have the code try to read the sun sensors, etc, even if there are no solar boards attached.

jbwillis replied:
Then report that the sun sensors failed the test.

yashikabatra15 replied:
It checks if the sun sensors are attached to the board / tries to read them earlier in the check. This check is built into the pycubed library, so when we initialize a satellite object we also initialize a hardware dictionary, which is what is printed earlier on in this file.

The input being asked for here feeds into a secondary hardware check, which will check for the functionality of the sun sensors, coil drivers, and burnwire ICs (which are all on the solar boards)

I can still remove this if you'd like though, as each individual file checks that the hardware exists before running the second-layer functionality test, and updates the results dictionary accordingly. All the results are printed out later in the test anyways, so it shouldn't be an issue.

jbwillis commented on state_machine/applications/system-check/main.py:
```diff
@@ -0,0 +1,115 @@
+    run_burnwire = True
+
+# ask to test logging
+logging_input = input("Would you like to test logging? (Y/N): ")
```
don't ask, always run

yashikabatra15 replied:
fixed!

jbwillis commented on state_machine/applications/system-check/main.py:
```diff
@@ -0,0 +1,115 @@
+    tests.logging_test.run(cubesat, hardware_dict, result_dict)
+
+# ask to test imu
+imu_input = input("Would you like to test the IMU sensors? (Y/N): ")
```
don't ask, always run

yashikabatra15 replied:
fixed

jbwillis commented on state_machine/applications/system-check/main.py:
```diff
@@ -0,0 +1,115 @@
+    tests.radio_test.run(cubesat, hardware_dict, result_dict, antenna_attached)
+
+# ask to test sun sensor
+if run_sun_sensor:
```
don't ask, always run

yashikabatra15 replied:
fixed!

jbwillis commented on state_machine/applications/system-check/main.py:
```diff
@@ -0,0 +1,115 @@
+        tests.sun_sensor_test.run(cubesat, hardware_dict, result_dict)
+
+# ask to test coil driver
+if run_coil_driver:
```
don't ask, always run

yashikabatra15 replied:
fixed

jbwillis commented on state_machine/applications/system-check/main.py:
```diff
@@ -0,0 +1,115 @@
+        tests.coil_test.run(cubesat, hardware_dict, result_dict)
+
+# ask to test burnwire
+if run_burnwire:
```
Ok to ask for this one

jbwillis commented on state_machine/applications/system-check/main.py:
```diff
@@ -0,0 +1,115 @@
+    tests.imu_test.run(cubesat, hardware_dict, result_dict)
+
+# ask to test radio
+radio_input = input("Would you like to test the radio? (Y/N): ")
```
Ok to ask for this one, but let's move it to the end.

yashikabatra15 replied:
fixed

jbwillis commented on state_machine/applications/system-check/tests/coil_test.py:
```diff
@@ -0,0 +1,137 @@
+    print("Testing Coil", str(coilidx), "for the following voltage levels: " +
+          str(projected_voltage1) + ", " + str(projected_voltage2 + ", " +
+          str(projected_voltage3)))
+    print("Please retrieve a magnetometer. A cellphone app will also work. " +
```
I think we should actually be able to do this test without a cell phone/external magnetometer by just reading the values of the IMU magnetometer before and during the test and taking the difference. That would be good enough to see that the coils are operating.

yashikabatra15 replied:
Implemented this, thanks for the tip! That definitely makes it easier on the user

jbwillis commented on state_machine/applications/system-check/tests/burnwire_test.py:
```diff
@@ -0,0 +1,121 @@
+    print(result_val_string)
+
+    # update dictionary values based on user test result values
+    if (abs(projected_voltage - actual_voltage) < 0.2):
```
Honestly, I have no idea how close these should actually be. Rather than predicting the voltage, you could just have the user indicate if the burn wire functions properly.

yashikabatra15 replied:
I remember when Neil and I tested it last, it was within 0.1 volts (which is why I expanded the range to 0.2). However, you are correct that this is something that's a little more unpredictable -- I can just ask for a Y/N input from the user and log the result of that if you prefer.

jbwillis commented on state_machine/applications/system-check/tests/burnwire_test.py:
```diff
@@ -0,0 +1,121 @@
+    return result_dict
+
+
+def run(cubesat, hardware_dict, result_dict):
```
I think there are some really good ideas behind this test but it is quite onerous, which makes it unlikely someone will actually run it. It also only captures the functionality of the burnwire circuit when no burn wire is attached. 

Let's simplify this to the following:
1. For each burn wire
2. Prompt the user to enter a voltage and time. If nothing is entered, use a default
3. prompt the user to provide input when they want the burn wire to turn on. suggest testing using either a multimeter or a piece of fishing line.
4. Ask if the test succeeded
5. Done

yashikabatra15 replied:
Done!

jbwillis commented on state_machine/applications/system-check/tests/burnwire_test.py:
```diff
@@ -0,0 +1,121 @@
+v2 = 0.5
+v3 = 0.75
+
+def user_test(burnnum, burnwire, vlevel):
```
See my comment on `run` for how to restructure this.

jbwillis commented on state_machine/applications/system-check/tests/coil_test.py:
```diff
@@ -0,0 +1,137 @@
+    return success
+
+
+def voltage_levelx(cubesat, result_dict, coilidx):
```
My general preference is to not abbreviate and use underscores to separate words. `dict` is probably ok, as it is a very common abbreviation and significantly shortens the word (though I still think `dictionary` is better), but I would use `coil_index` - it's much more readable than `coilidx`.

jbwillis commented on state_machine/applications/system-check/tests/coil_test.py:
```diff
@@ -0,0 +1,137 @@
+def voltage_levelx(cubesat, result_dict, coilidx):
+    """
+    All automation happens in this function
+    Select burnwire and voltage level
```
burnwire -> coil?

yashikabatra15 replied:
fixed

jbwillis commented on state_machine/applications/system-check/tests/coil_test.py:
```diff
@@ -0,0 +1,137 @@
+v2 = 0x15
+v3 = 0x1F
+
+def user_test(driver, coilidx):
```
`coilidx` -> `coil_index`. See my later comment.

yashikabatra15 replied:
fixed

jbwillis commented on state_machine/applications/system-check/tests/imu_test.py:
```diff
@@ -0,0 +1,147 @@
+        return acc, gyro, mag
+
+
+def imu_test(cubesat, result_dict, testidx):
```
testidx isn't really an index. maybe `test_type`

yashikabatra15 replied:
fixed

jbwillis commented on state_machine/applications/system-check/tests/imu_test.py:
```diff
@@ -0,0 +1,147 @@
+
+        # if change in magnetometer reading is more than 0, true
+        if mag > 0:
+            result_dict['IMU_Gyro_Turntable'] = (result_val_string, True)
```
Copy-paste error here, this shouldn't be `IMU_Gyro_Turntable`

yashikabatra15 replied:
fixed

jbwillis commented on state_machine/applications/system-check/tests/imu_test.py:
```diff
@@ -0,0 +1,147 @@
+        result_val_string = "Testing IMU when cubesat is rotating on a turntable. Gyro: " + str(gyro)
+
+        # if total gyro reading is within 2 deg/s^2 of angular velocity, return true
+        ang_vel = float(input("Please enter the speed the turntable was rotating at in deg/s: "))
```
Probably unrealistic for someone to know this number.

jbwillis replied:
Just have the person rotate it in their hand, and check that it has changed from when the cubesat was still.

jbwillis replied:
Eventually we'll use a turntable, but that won't fall under the umbrella of this script.

jbwillis commented on state_machine/applications/system-check/tests/logging_test.py:
```diff
@@ -0,0 +1,17 @@
+    if not hardware_dict['SDcard']:
+        result_dict['SD_Card_Logging'] = ('Cannot test logging; no SD Card detected', False)
+        return result_dict
+
```
Is this incomplete at the moment?

yashikabatra15 replied:
Yep, it's not complete at the moment because i was/am still ediitng the SD card functions. I'm mainly done with those though (they're on the misc repo), so I'll probably define a simple test before the next update. 

It'll likely just create a file, make sure it exists and has a timestamp, and then delete the file and make sure it's no longer there.

jbwillis commented on state_machine/applications/system-check/main.py:
```diff
@@ -0,0 +1,115 @@
+
+# print acknowledgement that test has started
+print("Running System Check...")
+print("Initialization has concluded. Printing results...")
```
Oh, I see that you're printing the hardware dictionary. Ok, maybe make your message a bit more clear. 

"Hardware successfully initialized: "
...

yashikabatra15 replied:
Fixed!

jbwillis reviewed (changes requested):
I made a bunch of comments and have more, but I think there are a few general "principles" that I'd like to see this system test script follow and that will resolve many of my comments.

Make it as automatic and simple as possible. People using this don't want to have to make lots of decisions (plus that introduces the opportunity for error). Any time you can do something automatically do it. This includes asking if a test should be run. Unless there is significant external hardware or possibility for damage to something if the test is run, just run it.

Other thought is to have this be a PR into `update-driver` rather than `main` so we are pulling in the changes from there as well.

thetazero commented:
> People using this don't want to have to make lots of decisions (plus that introduces the opportunity for error). Any time you can do something automatically do it. This includes asking if a test should be run. 

I like this general idea, but I think it would be super convenient to have the option to just run some particular test. I think this would be really helpful in certain cases such as: 
- Making a modification to the way the pycubed.py library interfaces with the IMU (only IMU tests should be run in this case)
- Running some particularly involved and time consuming test that doesn't run by default. Such as one that requires you to cover up/shine a light on each sun sensor one at a time. 

It may be better to address this in a separate pull request however. What do you think @yashikabatra15, @jbwillis

jbwillis commented:
Responding to @thetazero here are my thoughts.

I think what we are saying is we want to have a separation of the fast manual tests and the slow manual tests. 

Slow manual tests should be something that can be run separately, but fast manual tests are already fast, so there isn't a downside to running them frequently.

One objective I have is for us to have a script that we can run when a new board is made or when a board is added to the system and quickly see what is functioning. This is primarily answering the question "is it there?"

The other is to do a more complete test of individual subsystems.

So, I think we structure this in a way that everything fast + automatic runs every time. Then, you are given the option to run the slower tests that require human input.

yashikabatra15 commented on state_machine/applications/system-check/main.py:
```diff
@@ -0,0 +1,115 @@
+
+"""
+TODO:
+run through files and documents for last check
```
Yes - it was mainly to remind myself to add documentation and fix print statement formatting; I must've forgotten to remove it

yashikabatra15 commented on state_machine/applications/system-check/main.py:
```diff
@@ -0,0 +1,115 @@
+    'IMU_Temp': ('', False),
+    'Radio_Receive_Beacon': ('', False),
+    'Radio_Send_Beacon': ('', False),
+    'Sun_MinusY_Dark': ('', False),
```
Fixed!

yashika-batra commented:
Some notes on the system-check-script PR:
- Merged new driver and transmission code from the update-driver branch onto this branch
- Changed as per your comments
    - Burnwire test
    - Coil Driver test
    - Gyroscope test
- Added:
    - Simple IMU temperature sensor test
    - SD card test
    - I2C Scan: returns the addresses of devices connected to each I2C bus

Note on the structure of the tests: the base level check of “is it there” is run when the satellite object is initialized within pycubed. The results of this check populate the “hardware_dict” that’s referenced in main.py. Everyother file / test first checks the hardware_dict to ensure that the object being tested is in fact there, and then runs the higher level “is it functional” check.
- Every “is it functional” test, apart from radio and burnwire, is fairly fast (5 minutes total), so I took your suggestion on only explicitly asking the user if they want to run the radio and burnwire test
- Let me know if there’s any other user input changes y’all want to add (given this setup) / if there’s any structural changes you want me to make to change this
